File: main_project/client/java/OceAdapter/src/main/java/com/xiaonei/xce/xcealert/XceAlertAdapter.java
package com.xiaonei.xce.xcealert;

import java.util.List;
import java.util.Vector;

import com.xiaonei.xce.XceAlertPrx;
import com.xiaonei.xce.XceAlertPrxHelper;
import xce.clusterstate.ReplicatedClusterAdapter;
import com.xiaonei.xce.log.Oce;

public class XceAlertAdapter extends ReplicatedClusterAdapter {
	private static final String ctr_endpoints = "ControllerXceAlert";

	private static int _interval = 1;

	private static XceAlertAdapter alert_ = new XceAlertAdapter();

	public static XceAlertAdapter getInstance() {
		return alert_;
	}

	private XceAlertAdapter() {
		super(ctr_endpoints, _interval);
	}

	public void notifyAlert(List<String> ids, String msg) {
		XceAlertPrx prx = null;
		try {
			prx = getXceAlert();
			prx.notifyAlert(ids.toArray(new String[ids.size()]), msg);
		} catch (Exception e) {
			Oce.getLogger().error(this.getClass().getName() + ".notifyAlert, prx = " + prx, e);
		}
	}
	
	public void notifyAlertN(List<String> subjects, String msg, int sendType) {
		XceAlertPrx prx = null;
		try {
			prx = getXceAlert();
			//sendType must be 1, 2, or 3.
			//1 send msg to mobile; 2 send msg to email, 3 send msg to both
			prx.notifyAlertN(subjects.toArray(new String[subjects.size()]), msg, sendType);
		} catch (Exception e) {
			Oce.getLogger().error(this.getClass().getName() + ".notifyAlert, prx = " + prx, e);
		}
	}

	protected XceAlertPrx getXceAlert() {
		Ice.ObjectPrx prx0 = getValidProxy(0, 300);
		if (prx0 != null) {
			XceAlertPrx prx = XceAlertPrxHelper
					.uncheckedCast(prx0.ice_timeout(300).ice_twoway());
			return prx;
		} else {
			return null;
		}
	}

	protected XceAlertPrx getXceAlertOneway() {
		Ice.ObjectPrx prx0 = getValidProxy(0, 300);
		if (prx0 != null) {
			XceAlertPrx prx = XceAlertPrxHelper
					.uncheckedCast(prx0.ice_timeout(300).ice_twoway());
			return prx;
		} else {
			return null;
		}
	}
}

File: main_project/OceCxxAdapter/src/XceAlertAdapter.cpp
#include "XceAlertAdapter.h"

using com::xiaonei::xce::XceAlertAdapter;

XceAlertAdapter::XceAlertAdapter() : MyUtil::ReplicatedClusterAdapterI <XceAlertPrx>("ControllerXceAlert", 120, 300) {
}

void XceAlertAdapter::notifyAlert(const MyUtil::StrSeq& typeSeq, const std::string& msg) {
	XceAlertPrx prx = getProxy(0);
	if(prx)
	{
		try {
			getProxy(0)->notifyAlert(typeSeq, msg);
		} catch (const Ice::Exception& e) {
			MCE_WARN("XceAlertAdapter::notifyAlert Ice::Exception " << e);
		}
	}else
	{ 
		MCE_WARN("XceAlertAdapter::notifyAlert no prx");
	}
}

void XceAlertAdapter::notifyAlertN(const MyUtil::StrSeq& typeSeq, const std::string& msg, int sendType) {
	try {
		getProxy(0)->notifyAlertN(typeSeq, msg, sendType);
	} catch (const Ice::Exception& e) {
		MCE_WARN("XceAlertAdapter::notifyAlertN Ice::Exception " << e);
	}
}

void XceAlertAdapter::reloadAlertMap() {
	try {
		getProxy(0)->reloadAlertMap();
	} catch (const Ice::Exception& e) {
		MCE_WARN("XceAlertAdapter::reload Ice::Exception " << e);
	}
}	
